Compile errors with 0.21.1

### Bug Description

Updated to 0.21.1, now getting several of this error:

```
error[E0624]: associated constant `_PYO3_DEF` is private
  --> src/lib.rs:25:1
   |
25 | #[pymodule]
   | ^^^^^^^^^^^ private associated constant
```

This is using declarative modules. I assume it's related to https://github.com/PyO3/pyo3/pull/4009. Compiling with 0.21.0 works fine.

### Steps to Reproduce

```rust
mod whatever;

use pyo3::prelude::*;

#[pymodule]
mod rust_stuff {
    #[pymodule_export]
    use crate::whatever::SomePyClass;
}
```

### Backtrace

_No response_

### Your operating system and version

Ubuntu

### Your Python version (`python --version`)

3.12

### Your Rust version (`rustc --version`)

rustc 1.79.0-nightly (5f2c7d2bf 2024-03-25)

### Your PyO3 version

0.21.1

### How did you install python? Did you use a virtualenv?

...

### Additional Info

_No response_

If you add pub(crate) on that class, does it work?

On Tue, Apr 2, 2024, 2:11 PM Nick Drozd ***@***.***> wrote:

> Bug Description
>
> Updated to 0.21.1, now getting several of this error:
>
> error[E0624]: associated constant `_PYO3_DEF` is private
>   --> src/lib.rs:25:1
>    |
> 25 | #[pymodule]
>    | ^^^^^^^^^^^ private associated constant
>
> This is using declarative modules. I assume it's related to #4009
> <https://github.com/PyO3/pyo3/pull/4009>. Compiling with 0.21.0 works
> fine.
> Steps to Reproduce
>
> mod whatever;
>
> use pyo3::prelude::*;
>
> #[pymodule]
> mod rust_stuff {
> #[pymodule_export]
> use crate::whatever::SomePyClass;
> }
> Backtrace
>
> *No response*
> Your operating system and version
>
> Ubuntu
> Your Python version (python --version)
>
> 3.12
> Your Rust version (rustc --version)
>
> rustc 1.79.0-nightly (5f2c7d2bf 2024-03-25)
> Your PyO3 version
>
> 0.21.1
> How did you install python? Did you use a virtualenv?
>
> ...
> Additional Info
>
> *No response*
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PyO3/pyo3/issues/4036>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBE2VAV4CUMVR2V3R5LY3LYE5AVCNFSM6AAAAABFT3T2NWVHI2DSMVQWIX3LMV43ASLTON2WKOZSGIZDCMJTGIZTSNQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/newsfragments/4054.fixed.md
+++ b/newsfragments/4054.fixed.md
@@ -1,0 +1,1 @@
+Fixes a compile error when exporting a `#[pyclass]` living in a different Rust module using the declarative-module feature.

--- a/pyo3-macros-backend/src/pyclass.rs
+++ b/pyo3-macros-backend/src/pyclass.rs
@@ -1614,7 +1614,7 @@ impl<'a> PyClassImplsBuilder<'a> {
         quote! {
             impl #cls {
                 #[doc(hidden)]
-                const _PYO3_DEF: #pyo3_path::impl_::pymodule::AddClassToModule<Self> = #pyo3_path::impl_::pymodule::AddClassToModule::new();
+                pub const _PYO3_DEF: #pyo3_path::impl_::pymodule::AddClassToModule<Self> = #pyo3_path::impl_::pymodule::AddClassToModule::new();
             }
         }
     }

--- a/tests/test_declarative_module.rs
+++ b/tests/test_declarative_module.rs
@@ -9,6 +9,13 @@ use pyo3::types::PyBool;
 #[path = "../src/tests/common.rs"]
 mod common;
 
+mod some_module {
+    use pyo3::prelude::*;
+
+    #[pyclass]
+    pub struct SomePyClass;
+}
+
 #[pyclass]
 struct ValueClass {
     value: usize,
@@ -49,6 +56,10 @@ mod declarative_module {
     use super::*;
     #[pymodule_export]
     use super::{declarative_module2, double, MyError, ValueClass as Value};
+
+    // test for #4036
+    #[pymodule_export]
+    use super::some_module::SomePyClass;
 
     #[pymodule]
     mod inner {

